app/assets/javascripts/oxalis/model/sagas/save_saga.js

```diff
@@ -33,6 +34,40 @@ import { moveTreeComponent } from "oxalis/model/sagas/update_actions";
 
 const PUSH_THROTTLE_TIME = 30000; // 30s
 const SAVE_RETRY_WAITING_TIME = 5000;
+const UNDO_HISTORY_SIZE = 100;
+
+export function* collectUndoStates(): Generator<*, *, *> {
+  const undoStack = [];
+  const redoStack = [];
+
+  yield race({
+    initSkeleton: take("INITIALIZE_SKELETONTRACING"),
+    initVolume: take("INITIALIZE_VOLUMETRACING"),
+  });
+  let prevTracing = yield select(state => state.tracing);
+  while (true) {
+    const { userAction, undo } = yield race({
+      userAction: take([...SkeletonTracingSaveRelevantActions]),
+      undo: take("UNDO"),
+      redo: take("REDO"),
+    });
+    if (userAction) {
+      // Clear the redo stack when a new action is executed
+      redoStack.splice(0);
+      undoStack.push(prevTracing);
+      if (undoStack.length > UNDO_HISTORY_SIZE) undoStack.shift();
+    } else if (undo && undoStack.length) {
+      redoStack.push(prevTracing);
+      const newTracing = undoStack.pop();
+      yield put(setTracingAction(newTracing));
+    } else if (redoStack.length) {
```

Comment:
Shouldn't this be `redo && redoStack.length`? Otherwise, if my undoStack is empty and I try to do another undo, this code will execute a redo?